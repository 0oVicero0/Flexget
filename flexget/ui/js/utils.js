'use strict';

String.prototype.startsWith = function(prefix) {
  return this.indexOf(prefix) === 0;
};

function registerFlexModule(module) {
  angular.module('flexgetApp').requires.push(module.name);
}

app.menu = [];
app.routes = [];

function register_menu(href, caption, icon, order) {
  href = '/ui/#' + href;
  app.menu.push({href: href, caption: caption, icon: icon, order: order})
}

function register_route(name, url, controller, template) {
  app.routes.push({
    name: name,
    url: url,
    controller: controller,
    template: template
  });
}