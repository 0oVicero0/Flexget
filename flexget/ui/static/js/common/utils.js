'use strict';

String.prototype.startsWith = function(prefix) {
    return this.indexOf(prefix) === 0;
};

function registerFlexModule(module) {
  angular.module('flexgetApp').requires.push(module.name);
}