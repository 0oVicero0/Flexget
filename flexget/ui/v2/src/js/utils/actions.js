export const createAction = namespace => (type, payload, {
  ignore = false,
  message,
} = {}) => ({
  type,
  payload,
  error: (payload instanceof Error),
  meta: {
    ignore,
    message,
    namespace,
  },
});

export const createLoading = namespace => type => ({
  type,
  meta: {
    namespace,
    loading: true,
  },
});
