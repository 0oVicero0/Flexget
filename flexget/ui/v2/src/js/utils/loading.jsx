import Loadable from 'react-loadable';
import Loading from 'components/common/Loader';

export function createAsyncComponent(loader) {
  return Loadable({
    loading: Loading,
    loader,
  });
}

