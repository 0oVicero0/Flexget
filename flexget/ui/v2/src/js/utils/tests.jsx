import React from 'react';
import { MuiThemeProvider } from 'material-ui/styles';
import { MemoryRouter } from 'react-router-dom';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import theme from 'theme';

const mockStore = configureMockStore([thunk]);

export function themed(component) {
  return (
    <MuiThemeProvider theme={theme} >
      { component }
    </MuiThemeProvider>
  );
}

export function router(component) {
  return (
    <MemoryRouter>
      { component }
    </MemoryRouter>
  );
}

export function provider(component, state = {}) {
  return (
    <Provider store={mockStore(state)}>
      { component }
    </Provider>
  );
}
