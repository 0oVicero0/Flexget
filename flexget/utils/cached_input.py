import copy
import logging
from flexget.plugin import register_plugin

log = logging.getLogger('cached')


class cached(object):
    """
    Implements transparent caching decorator @cached for inputs.

    Decorator has two parameters

    :name: in which the configuration is present in feeds configuration.
    :key: in which the configuration has the cached resource identifier (ie. url). If the :key: is not
    given or present in the configuration :name: is expected to be a cache name (ie. url)

    Configuration assumptions may make this unusable in some (future) inputs
    """

    cache = {}

    def __init__(self, name, key=None):
        self.name = name
        self.key = key

    def __call__(self, func):

        def wrapped_func(*args, **kwargs):
            # get feed from method parameters
            feed = args[1]
            
            # detect api version
            api_ver = 1
            if len(args) == 3:
                api_ver = 2
            
            if api_ver == 1:
                # get name for a cache from feeds configuration
                if not self.name in feed.config:
                    raise Exception('@cache config name %s is not configured in feed %s' % (self.name, feed.name))
                config = feed.config[self.name]
            else:
                config = args[2]

            log.log(5, 'config: %s' % config)
            log.log(5, 'self.name: %s' % self.name)
            log.log(5, 'self.key: %s' % self.key)
            
            if api_ver == 1:
                if isinstance(config, dict) and self.key in config:
                    name = feed.config[self.name][self.key]
                else:
                    name = feed.config[self.name]
            else:
                if isinstance(config, dict) and self.key in config:
                    name = config[self.key]
                else:
                    name = config

            log.debug('cache name: %s (has: %s)' % (name, ', '.join(self.cache.keys())))

            if name in self.cache:
                # return from the cache
                log.log(5, 'cache hit')
                count = 0
                entries = []
                for entry in self.cache[name]:
                    fresh = copy.deepcopy(entry)
                    entries.append(fresh)
                count += 1
                if count > 0:
                    feed.verbose_progress('Restored %s entries from cache' % count, log)
                return entries
            else:
                log.log(5, 'cache miss')
                # call input event
                response = func(*args, **kwargs)
                # store results to cache
                if api_ver == 1:
                    log.debug('v1 storing to cache %s %s entries' % (name, len(feed.entries)))
                    try:
                        self.cache[name] = copy.deepcopy(feed.entries)
                    except TypeError:
                        # might be caused because of backlog restoring some idiotic stuff, so not neccessarily a bug
                        log.critical('Unable to save feed content into cache, if problem persists longer than a day please report this as a bug')
                else:
                    log.debug('storing to cache %s %s entries' % (name, len(feed.entries)))
                    try:
                        self.cache[name] = copy.deepcopy(response)
                    except TypeError:
                        # might be caused because of backlog restoring some idiotic stuff, so not neccessarily a bug
                        log.critical('Unable to save feed content into cache, if problem persists longer than a day please report this as a bug')
                return response
                    

        return wrapped_func


class CacheClearer(object):

    def on_process_start(self, feed):
        """Internal. Clears the input cache on every process start.
        This is neccessary for webui or otherwise it will only use cache.
        """
        log.debug('clearing cache')
        cached.cache = {}

register_plugin(CacheClearer, 'cache_clearer', builtin=True)
