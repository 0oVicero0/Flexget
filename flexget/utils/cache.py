import hashlib
from pathlib import Path

import requests

from flexget.utils.tools import log


def cached_resource(url, base_dir, force=False, max_size=250, directory='cached_resources'):
    """
    Caches a remote resource to local filesystem. Return a tuple of local file name and mime type, use primarily
    for API/WebUI.

    :param url: Resource URL
    :param force: Does not check for existence of cached resource, fetches the remote URL, ignores directory size limit
    :param max_size: Maximum allowed size of directory, in MB.
    :param directory: Name of directory to use. Default is `cached_resources`
    :return: Tuple of file path and mime type
    """
    mime_type = None
    hashed_name = hashlib.md5(url.encode('utf-8')).hexdigest()
    file_path = Path(base_dir) / directory / hashed_name
    directory = file_path.parent

    if not file_path.exists() or force:
        log.debug('caching %s', url)
        response = requests.get(url)
        response.raise_for_status()
        mime_type = response.headers.get('content-type')
        content = response.content
        directory.mkdir(parents=True, exist_ok=True)

        # Checks directory size and trims if necessary.
        size = dir_size(directory) / (1_024 * 1_024.0)
        if not force:
            while size >= max_size:
                log.debug('directory %s size is over the allowed limit of %s, trimming', size, max_size)
                trim_dir(directory)
                size = dir_size(directory) / (1_024 * 1_024.0)
        file_path.write_bytes(content)

    return str(file_path), mime_type


def dir_size(directory: Path):
    """
    Sums the size of all files in a given dir. Not recursive.

    :param directory: Directory to check
    :return: Summed size of all files in Bytes.
    """
    return sum(file.stat().st_size for file in directory.iterdir())


def trim_dir(directory: Path):
    """
    Removed the least accessed file on a given dir

    :param directory: Directory to check
    """

    def access_time(file: Path):
        return file.stat().st_atime

    files = sorted(directory.iterdir(), key=access_time)
    file_name = directory / files[0]
    log.debug('removing least accessed file: %s', file_name)
    file_name.unlink()
