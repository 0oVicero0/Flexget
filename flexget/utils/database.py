from datetime import datetime
from functools import update_wrapper
from sqlalchemy import extract, func, case
from sqlalchemy.orm import synonym
from flexget.manager import Session
from flexget.utils import qualities


def with_session(func):
    """"Creates a session if one was not passed via keyword argument to the function"""

    def wrapper(*args, **kwargs):
        if not kwargs.get('session'):
            kwargs['session'] = Session(autoflush=True)
            try:
                return func(*args, **kwargs)
            finally:
                kwargs['session'].close()
        else:
            return func(*args, **kwargs)
    return wrapper


def pipe_list_synonym(name):
    """Converts pipe separated text into a list"""

    def getter(self):
        attr = getattr(self, name)
        if attr:
            return attr.strip('|').split('|')

    def setter(self, value):
        if isinstance(value, basestring):
            setattr(self, name, value)
        else:
            setattr(self, name, '|'.join(value))

    return synonym(name, descriptor=property(getter, setter))


def text_date_synonym(name):
    """Converts Y-M-D date strings into datetime objects"""

    def getter(self):
        return getattr(self, name)

    def setter(self, value):
        if isinstance(value, basestring):
            setattr(self, name, datetime.strptime(value, '%Y-%m-%d'))
        else:
            setattr(self, name, value)

    return synonym(name, descriptor=property(getter, setter))


def safe_pickle_synonym(name):
    """Used to store Entry instances into a PickleType column in the database.

    In order to ensure everything can be loaded after code changes, makes sure no custom python classes are pickled.
    """

    def only_builtins(item):
        """Casts all subclasses of builtin types to their builtin python type. Works recursively on iterables.

        Raises ValueError if passed an object that doesn't subclass a builtin type.
        """

        supported_types = [str, unicode, int, float, bool, datetime]
        # dict, list, tuple and set are also supported, but handled separately

        if type(item) in supported_types:
            return item
        elif isinstance(item, dict):
            result = {}
            for key, value in item.iteritems():
                try:
                    result[key] = only_builtins(value)
                except TypeError:
                    continue
            return result
        elif isinstance(item, (list, tuple, set)):
            result = []
            for value in item:
                try:
                    result.append(only_builtins(value))
                except ValueError:
                    continue
            if isinstance(item, list):
                return result
            elif isinstance(item, tuple):
                return tuple(result)
            else:
                return set(result)
        else:
            for s_type in supported_types:
                if isinstance(item, s_type):
                    return s_type(item)

        # If item isn't a subclass of a builtin python type, raise ValueError.
        raise TypeError('%r is not a subclass of a builtin python type.' % type(item))

    def getter(self):
        return getattr(self, name)

    def setter(self, entry):
        setattr(self, name, only_builtins(entry))

    return synonym(name, descriptor=property(getter, setter))


def quality_synonym(name):
    """Get and set property as Quality objects, store as text in the database."""

    def getter(self):
        return qualities.get(getattr(self, name))

    def setter(self, value):
        if isinstance(value, basestring):
            setattr(self, name, value)
        else:
            setattr(self, name, value.name)

    return synonym(name, descriptor=property(getter, setter))


class property_(object):
    """A property for an sqlalchemy model that can also return the proper sql to be used in queries."""

    def __init__(self, fget, fset=None, fdel=None, expr=None):
        self.fget = fget
        self.fset = fset
        self.fdel = fdel
        self.expr = expr or fget
        update_wrapper(self, fget)

    def __get__(self, instance, owner):
        if instance is None:
            return self.expr(owner)
        else:
            return self.fget(instance)

    def __set__(self, instance, value):
        self.fset(instance, value)

    def __delete__(self, instance):
        self.fdel(instance)

    def setter(self, fset):
        self.fset = fset
        return self

    def deleter(self, fdel):
        self.fdel = fdel
        return self

    def expression(self, expr):
        self.expr = expr
        return self


def year_property(date_attr):

    def getter(self):
        date = getattr(self, date_attr)
        return date and date.year

    def expr(cls):
        return extract('year', getattr(cls, date_attr))

    return property_(getter, expr=expr)


def lower_property(str_attr):
    """A property based on a text column that returns or queries on the lower case version of the string."""

    def getter(self):
        string = getattr(self, str_attr)
        return string and string.lower()

    def expr(cls):
        return func.lower(getattr(cls, str_attr))

    return property_(getter, expr=expr)


def quality_comp_property(str_attr):
    """Returns the value for the quality stored, which can be filtered against or used to sort queries."""

    def getter(self):
        return qualities.get(getattr(self, str_attr)).value

    def expr(cls):
        whens = {}
        for qual in qualities.all():
            whens[qual.name] = qual.value
        return case(value=getattr(cls, str_attr), whens=whens)

    return property_(getter, expr=expr)
