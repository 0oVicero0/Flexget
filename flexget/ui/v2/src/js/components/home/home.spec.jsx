import React from 'react';
import renderer from 'react-test-renderer';
import HomePage from 'components/home';
import { themed } from 'utils/tests';

describe('components/home', () => {
  it('renders correctly', () => {
    const tree = renderer.create(
      themed(<HomePage />),
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
