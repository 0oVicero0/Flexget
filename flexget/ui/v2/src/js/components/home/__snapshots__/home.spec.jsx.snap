// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/home renders correctly 1`] = `
<div
  className="MuiPaper-root-6 MuiPaper-shadow2-a MuiPaper-rounded-7 MuiCard-root-5 InfoCard-card-1"
>
  <div
    className="MuiCardContent-root-10 MuiCardHeader-root-x InfoCard-cardHeader-2"
  >
    <div
      className="MuiCardHeader-content-z"
    >
      <span
        className="MuiTypography-root-11 MuiTypography-headline-16"
      >
        Flexget Web Interface
      </span>
      <span
        className="MuiTypography-root-11 MuiTypography-body1-1a MuiTypography-colorSecondary-1l"
      >
        Under Development
      </span>
    </div>
  </div>
  <div
    className="MuiCardContent-root-10"
  >
    <p
      className="InfoCard-boldText-3"
    >
      We need your help! If you are a React developer or can help with the layout/design/css then please join in the effort!
    </p>
    <p>
      The interface is not yet ready for end users. Consider this preview only state.
    </p>
    <p>
      If you still use it anyways, please report back to us on how well it works, issues, ideas etc...
    </p>
    <p>
      There is a functional API with documentation available at 
      <a
        href="/api"
      >
        /api
      </a>
    </p>
    <p>
      More information: 
      <a
        href="http://flexget.com/wiki/Web-UI/v2"
        rel="noopener noreferrer"
        target="_blank"
      >
        http://flexget.com/wiki/Web-UI/v2
      </a>
    </p>
    <p>
      Gitter Chat: 
      <a
        href="https://gitter.im/Flexget/Flexget"
        rel="noopener noreferrer"
        target="_blank"
      >
        https://gitter.im/Flexget/Flexget
      </a>
    </p>
  </div>
  <div
    className="MuiCardActions-root-1n InfoCard-cardActions-4"
  >
    <a
      aria-label="Github"
      className="MuiButtonBase-root-1y MuiIconButton-root-1p MuiCardActions-actionSpacing-1o"
      href="https://github.com/Flexget/Flexget"
      onBlur={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchStart={[Function]}
      rel="noopener noreferrer"
      tabIndex="0"
      target="_blank"
    >
      <span
        className="MuiIconButton-label-1v"
      >
        <span
          aria-hidden="true"
          className="material-icons MuiIcon-root-20 MuiIconButton-icon-1w fa fa-github"
        />
      </span>
      <span
        className="MuiTouchRipple-root-27"
      />
    </a>
    <a
      aria-label="Flexget.com"
      className="MuiButtonBase-root-1y MuiIconButton-root-1p MuiCardActions-actionSpacing-1o"
      href="https://flexget.com"
      onBlur={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchStart={[Function]}
      rel="noopener noreferrer"
      tabIndex="0"
      target="_blank"
    >
      <span
        className="MuiIconButton-label-1v"
      >
        <span
          aria-hidden="true"
          className="material-icons MuiIcon-root-20 MuiIconButton-icon-1w fa fa-home"
        />
      </span>
      <span
        className="MuiTouchRipple-root-27"
      />
    </a>
    <a
      aria-label="Gitter"
      className="MuiButtonBase-root-1y MuiIconButton-root-1p MuiCardActions-actionSpacing-1o"
      href="https://gitter.im/Flexget/Flexget"
      onBlur={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchStart={[Function]}
      rel="noopener noreferrer"
      tabIndex="0"
      target="_blank"
    >
      <span
        className="MuiIconButton-label-1v"
      >
        <span
          aria-hidden="true"
          className="material-icons MuiIcon-root-20 MuiIconButton-icon-1w fa fa-comment"
        />
      </span>
      <span
        className="MuiTouchRipple-root-27"
      />
    </a>
    <a
      aria-label="Forum"
      className="MuiButtonBase-root-1y MuiIconButton-root-1p MuiCardActions-actionSpacing-1o"
      href="https://discuss.flexget.com"
      onBlur={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchStart={[Function]}
      rel="noopener noreferrer"
      tabIndex="0"
      target="_blank"
    >
      <span
        className="MuiIconButton-label-1v"
      >
        <span
          aria-hidden="true"
          className="material-icons MuiIcon-root-20 MuiIconButton-icon-1w fa fa-forumbee"
        />
      </span>
      <span
        className="MuiTouchRipple-root-27"
      />
    </a>
  </div>
</div>
`;
